eventstore/store.go

```diff
@@ -1,49 +1,188 @@
+func NewStore(opts ...StoreOption) (*Store, error) {
+	config := &StoreConfig{}
+	for _, opt := range opts {
+		opt(config)
+	}
+	if config.Datastore == nil {
+		datastore, err := newDefaultDatastore(config.RepoPath)
+		if err != nil {
+			return nil, err
+		}
+		config.Datastore = datastore
+	}
+	if config.EventCodec == nil {
+		config.EventCodec = newDefaultEventCodec()
 	}
+	logWriter := &lumberjack.Logger{
+		Filename:   filepath.Join(config.RepoPath, defaultRepoPath, "log"),
+		MaxSize:    10, // megabytes
+		MaxBackups: 3,
+		MaxAge:     30, // days
+	}
+	if config.Debug {
+		err := setLogLevels(map[string]logger.Level{
+			"store":       logger.DEBUG,
+			"threads":     logger.DEBUG,
+			"threadstore": logger.DEBUG,
+		}, logWriter, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ownThreadService := false
+	if config.Threadservice == nil {
+		ownThreadService = true
+		ts, err := newDefaultThreadservice(ctx, logWriter, config.ListenPort, config.RepoPath, config.Debug)
+		if err != nil {
+			cancel()
+			return nil, err
+		}
+		config.Threadservice = ts
+	}
+
+	s := &Store{
+		ctx:              ctx,
+		cancel:           cancel,
+		datastore:        config.Datastore,
+		dispatcher:       newDispatcher(config.Datastore),
+		eventcodec:       config.EventCodec,
+		models:           make(map[reflect.Type]*Model),
+		localEventsBus:   &localEventsBus{bus: broadcast.NewBroadcaster(0)}, // ToDo: discuss about buffered chan, and implication on eventstore & thread sync tradeoffs
+		stateChanged:     &stateChangedNotifee{bus: broadcast.NewBroadcaster(1)},
+		threadservice:    config.Threadservice,
+		ownThreadService: ownThreadService,
+	}
+	return s, nil
+}
+
+func (s *Store) ThreadID() (thread.ID, bool, error) {
+	v, err := s.datastore.Get(dsStoreThreadID)
+	if err == ds.ErrNotFound {
+		return thread.ID{}, false, nil
+	}
+	if err != nil {
+		return thread.ID{}, false, err
+	}
+	id, err := thread.Cast(v)
+	return id, true, err
+}
+
+// Start should be called immediatelly after registering all schemas and before
+// any operation on them. If the store already boostraped on a thread, it will
+// continue using that thread. In the opposite case, it will create a new thread.
+func (s *Store) Start() error {
+	id, found, err := s.ThreadID()
+	if err != nil {
+		return err
+	}
+	if !found {
+		id = thread.NewIDV1(thread.Raw, 32)
+		if _, err := util.CreateThread(s.threadservice, id); err != nil {
+			return err
+		}
+		if err := s.datastore.Put(dsStoreThreadID, id.Bytes()); err != nil {
+			return err
+		}
+	}
+	adapter := newSingleThreadAdapter(s.ctx, s, id)
+	adapter.Start()
+	return nil
+}
+
+// StartFromAddr should be called immediatelly after registering all schemas
+// and before any operation on them. It pulls the current Store thread from
+// thread addr
+func (s *Store) StartFromAddr(addr ma.Multiaddr, followKey, readKey *symmetric.Key) error {
```

Comment:
Similarly with `Start`, `StartFromAddr` will bootstrap a `Store` from a Thread addr.
If other peer send us a invitation link, using that info we can start syncing things. 

As both method comments mention, models should register all types before calling start. If that isn't the case, when an Event will be reducer, no model would be found and would fail. I think this can be eventually solved by registering schema data in the log itself, so to make this thing automatic (and kind of consistent too). 

Anyway, doesn't seem too bad to ask to register all things before starting things (since that's how many similar things work). But seems like a good idea for the future. (Not to say how schema updates should be handled? :headexplode: :stuck_out_tongue_closed_eyes: )

Reply:
Ahh, I see. You answered the question above. 

`StartFromAddr`, I like that - makes it feel very easy to get going.

Reply:
Related to your thought about the schema... we could add an additional address protocol like,
```
/thread/<threadID>/model/<entityID>
```
Then, either check the address for the model protocol or add an additional method, `StartFromEntity`.

This would fit naturally with "getting" an entity like a shared google doc, etc. and would also mean you could return the schema associated with that model. Perhaps a you would ignore events not associated with that entity, perhaps not.

Reply:
> Ahh, I see. You answered the question above.
> 
> `StartFromAddr`, I like that - makes it feel very easy to get going.

Yes!, that was the idea. Hopefully will become more clear when making the more serious app on how it feel to use this. The `e2e_counter` example is quite dry. 



> Related to your thought about the schema... we could add an additional address protocol like,
> 
> ```
> /thread/<threadID>/model/<entityID>
> ```
> 
> Then, either check the address for the model protocol or add an additional method, `StartFromEntity`.
> 
> This would fit naturally with "getting" an entity like a shared google doc, etc. and would also mean you could return the schema associated with that model. Perhaps a you would ignore events not associated with that entity, perhaps not.

Interesting. Might be an interesting feature... only following some models.